rpc, p2p: add `addpermissionflags` RPC and allow whitelisting outbound

brunoerg opened the pull request:
Built this PR on top of #17167 (that's been closed due to inactivity but had some Concept ACK). So, it allows whitelisting outbound peers.

This PR adds a new RPC `addpermissionflags` to be able to set up permission flags -`whitelist` thru RPC, so we don't need to restart our node if we want to add new flags.

E.g.
```sh
$ ./src/bitcoin-cli addpermissionflags ["noban", "mempool", "in", "out"] "127.0.0.1"
```

luke-jr committed:
net: Move PF_ISIMPLICIT interpretation from AcceptConnection to new InitializePermissionFlags

luke-jr committed:
net: Move extra service flag into InitializePermissionFlags

luke-jr committed:
Accept "in" and "out" flags to -whitelist to allow whitelisting outgoing connections

brunoerg committed:
p2p, rpc: add `addpermissionflags`

brunoerg committed:
net: make `vWhitelistedRange` and `vWhitelistedRangeOutgoing` public

brunoerg committed:
test: add coverage for `addpermissionflags` RPC

brunoerg committed:
doc: add release notes for `addpermissionflags`

luke-jr reviewed (changes requested):
I think it would be better to specify as `[String "IP" or Number nodeid, [list of permissions], [list of permissions to remove]]` or similar.

MarcoFalke commented on src/rpc/net.cpp:
```diff
@@ -272,6 +272,49 @@ static RPCHelpMan getpeerinfo()
 
+static RPCHelpMan addwhitelist()
+{
+    return RPCHelpMan{"addwhitelist",
```
Wouldn't it be better to call this RPC `addpermissionflags`, to better convey that this is setting flags and not a single black/white bool?

ajtowns replied:
+1 -- we also have an rpc whitelist, so making it clear this changes net permission flags, not the rpc whitelist itself seems good

MarcoFalke commented on src/net.cpp:
```diff
@@ -575,12 +581,46 @@ void CNode::CloseSocketDisconnect()
+    }
+}
+
+std::string ConnectionTypeAsString(ConnectionType conn_type)
```
Is this used for anything?

brunoerg replied:
Seems not, gonna remove it. It seems a copy/paste from `CNode`

stickies-v commented on src/rpc/net.cpp:
```diff
@@ -272,6 +272,49 @@ static RPCHelpMan getpeerinfo()
+    const auto permissions{request.params[0].get_array()};
+
+    for (size_t i = 0; i < permissions.size(); i++) {
+        NetWhitelistPermissions subnet;
```
nit: `m_subnet` is a member of `NetWhitelistPermissions`, isn't `permissions` a more accurate name?
```suggestion
        NetWhitelistPermissions permissions;
```

brunoerg replied:
Agreed.

stickies-v commented on src/rpc/net.cpp:
```diff
@@ -272,6 +272,49 @@ static RPCHelpMan getpeerinfo()
+        NetWhitelistPermissions subnet;
+        ConnectionDirection connection_direction;
+        bilingual_str error;
+        if (!NetWhitelistPermissions::TryParse(permissions[i].get_str(), subnet, connection_direction, error)) throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid permissions");
```
nit: line length
```suggestion
        if (!NetWhitelistPermissions::TryParse(permissions[i].get_str(), subnet, connection_direction, error)) {
            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid permissions");
        }
```

stickies-v commented on src/rpc/net.cpp:
```diff
@@ -272,6 +272,49 @@ static RPCHelpMan getpeerinfo()
+
+    const auto permissions{request.params[0].get_array()};
+
+    for (size_t i = 0; i < permissions.size(); i++) {
```
```suggestion
    for (const auto& permission : permissions) {
```

stickies-v commented on src/rpc/net.cpp:
```diff
@@ -272,6 +272,49 @@ static RPCHelpMan getpeerinfo()
     };
 }
 
+static RPCHelpMan addwhitelist()
```
A question rather than a suggestion: should the docs highlight that these permissions do not persist beyond a node reboot? Intuitively, I'd have expected this kind of effect to be persistent, so maybe highlighting that is useful if I'm not the only one?

brunoerg replied:
It's intuitive for me these permissions aren't persistent since there is no way to remove them (thinking about a follow-up adding an option for removal in this RPC), but sounds goods for me to add it in the docs to clarify.

stickies-v commented on src/rpc/net.cpp:
```diff
@@ -272,6 +272,49 @@ static RPCHelpMan getpeerinfo()
+    return RPCHelpMan{"addwhitelist",
+                "\nAdd permission flags to the peers using the given IP address (e.g. 1.2.3.4) or\n"
+                "CIDR-notated network (e.g. 1.2.3.0/24). Uses the same permissions as -whitebind.\n"
+                "Additional flags in and out control whether permissions apply to incoming connections and/or outgoing (default: incoming only).\n",
```
Perhaps also useful to include this in the example?
```suggestion
                "Additional flags \"in\" and \"out\" control whether permissions apply to incoming and/or outgoing connections (default: incoming only).\n",
```

stickies-v commented on src/net_permissions.cpp:
```diff
@@ -104,11 +117,11 @@ bool NetWhitebindPermissions::TryParse(const std::string& str, NetWhitebindPermi
 }
 
-bool NetWhitelistPermissions::TryParse(const std::string& str, NetWhitelistPermissions& output, bilingual_str& error)
+bool NetWhitelistPermissions::TryParse(const std::string& str, NetWhitelistPermissions& output, ConnectionDirection& output_connection_direction, bilingual_str& error)
```
Not a big fan of adding extra out-parameters, I think `util::Result` would be more appropriate here (but also a bit of an overhaul so wouldn't be opposed to doing it in follow-up)

stickies-v reviewed (commented):
Concept ACK-ish, in that whitelisting shouldn't require a node restart.

Where I'm not 100% sure (as [commented](https://github.com/bitcoin/bitcoin/pull/26441#discussion_r1015627587)), is that this conceptually feels much more like a persistent setting instead of the ephemeral nature (i.e. effect gone after node restart) that these RPC commands have. Could be a source of confusion?

brunoerg commented:
I changed it to draft to work on some stuff some reviewers suggested here. Gonna advertise when it ready for review.

brunoerg commented:
Force-pushed changing the approach and addressing reviews! 

- Instead of using the same format from `-whitelist=` (e.g. `permissions@ip address or network`), now it has 2 params - an array of flags and an ip address or network.
  E.g.
  ```sh
  ./src/bitcoin-cli addpermissionflags ["noban", "mempool", "in", "out"] "127.0.0.1"
  ```
- Addressed https://github.com/bitcoin/bitcoin/pull/26441#discussion_r1015155904 by changing this RPC to `addpermissionflags`
- Addressed other minor changes

codo1 commented on src/rpc/net.cpp:
```diff
@@ -272,6 +272,61 @@ static RPCHelpMan getpeerinfo()
+                RPCResult{RPCResult::Type::NONE, "", ""},
+                RPCExamples{
+                    HelpExampleCli("addpermissionflags", "[\"noban\", \"mempool\", \"in\", \"out\"] \"127.0.0.1\"")
+            + HelpExampleRpc("addpermissionflags", "[\"noban\", \"mempool\", \"in\", \"out\"], \"127.0.0.1\" }")
```
Nit, probably only fix if file is touched anyway: one closing bracket "}" too much near the end of the line.

pablomartin4btc reviewed (commented):
Approach ACK, run automated tests (unit, functional & fuzz), pending of manual testing.

Regarding whitelisting outbound connections found some comments from @gmaxwell & @laanwj [here](https://github.com/bitcoin/bitcoin/pull/10051).

Please add "`Review club`" label ([*](https://bitcoincore.reviews/26441)).

LarryRuane commented on src/net.cpp:
```diff
@@ -949,14 +959,8 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
-        NetPermissions::AddFlag(permission_flags, NetPermissionFlags::Mempool);
-        NetPermissions::AddFlag(permission_flags, NetPermissionFlags::NoBan);
-    }
+    AddWhitelistPermissionFlags(permission_flags, addr, vWhitelistedRange);
```
The first commit ("net: Move PF_ISIMPLICIT interpretation from AcceptConnection to  ...") doesn't compile; this call to `AddWhitelistPermissionFlags()` should not have the `vWhitelistedRange` until the third commit, "Accept "in" and "out" flags to -whitelist...".

brunoerg replied:
Done in #27114

LarryRuane commented on src/net.cpp:
```diff
@@ -555,7 +560,8 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
                              /*inbound_onion=*/false,
-                             CNodeOptions{ .i2p_sam_session = std::move(i2p_transient_session) });
+                             CNodeOptions{ .i2p_sam_session = std::move(i2p_transient_session),
+                                           .permission_flags = permission_flags });
```
I got a compile error with the third commit, "Accept "in" and "out" flags to -whitelist..." because these are out of order, which you fixed later, but each commit should compile.

brunoerg replied:
Done in #27114

LarryRuane reviewed (changes requested):
I'll review more soon, but here are a couple of instances where changes are not in the correct commits. Each commit should compile (and all tests should pass, but I didn't check that).

hernanmarino reviewed (commented):
tested ACK. This might help with a couple of functional tests.

brunoerg commented:
Converted it to draft to address suggestions and re-build it on top of #27114.

Riahiamirreza commented on src/net_permissions.cpp:
```diff
@@ -85,7 +98,7 @@ bool NetWhitebindPermissions::TryParse(const std::string& str, NetWhitebindPermi
     NetPermissionFlags flags;
     size_t offset;
-    if (!TryParsePermissionFlags(str, flags, offset, error)) return false;
+    if (!TryParsePermissionFlags(str, flags, nullptr, offset, error)) return false;
```
nit: Why not using `ConnectionDirection::None` rather than `nullptr` to make it more clear?

brunoerg commented:
Closing it to focus on other approach.
